File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temp: List[int]) -> List[int]:
        li = list()
        n = len(temp)
        ans = [0 for x in range(n)]

        for i in range(n-1, -1, -1):
            if len(li)==0:
                ans[i] = 0
            else:
                while len(li)>0 and li[-1][0] <= temp[i]:
                    li.pop()
                if len(li)==0:
                    ans[i] = 0
                else:
                    ans[i] = li[-1][1]-i
            li.append([temp[i], i])
        return ans
        

File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temp) {
        int n = temp.size();
        vector<int>ans(n);
        stack<pair<int,int>>st;
        for(int i = n-1; i>=0; i--){
            while(st.size()>0 and st.top().first<=temp[i]) st.pop();
            if(st.size()==0) ans[i] = 0;
            else ans[i] = st.top().second-i;
            st.push({temp[i], i});
        }
        return ans;
    }
};